sql/mysqld.cc

```diff
@@ -3195,11 +3210,16 @@ bool one_thread_per_connection_end(THD *thd, bool block_pthread)
   PSI_THREAD_CALL(delete_current_thread)();
 #endif
 
-  // Decrement connection count while holding LOCK_thread_count
+  mysql_mutex_lock(&LOCK_thread_count);
+  if (!global_thread_count.load(std::memory_order_relaxed)) {
+    mysql_cond_broadcast(&COND_thread_count);
+  }
+  dec_connection_count_locked();
+
   if (block_pthread)
-    block_pthread= block_until_new_connection();
+    block_pthread= block_until_new_connection_halflock();
```

Comment:
Nit - I personally don't like having a function release a lock that it did not acquire.  I think it leads to less readable code.  Can we change `block_until_new_connection_halflock()` to not release the lock and this code to the following?

```
if (block_pthread)
  block_pthread= block_until_new_connection();

mysql_mutex_unlock(&LOCK_thread_count);
```